libmultilabel/nn/cluster.py

```diff
@@ -0,0 +1,121 @@
+        output_dir: directory to store the clustering file
+    """
+    # skip constructing label tree if the output file already exists
+    output_dir = output_dir if isinstance(output_dir, Path) else Path(output_dir)
+    cluster_path = output_dir / f"{CLUSTER_NAME}{FILE_EXTENSION}"
+    if cluster_path.exists():
+        logger.info("Clustering has finished in a previous run")
+        return
+
+    # meta info
+    logger.info("Label clustering started")
+    logger.info(f"Cluster size: {cluster_size}")
+    num_labels = sparse_y.shape[1]
+    # The height of the tree satisfies the following inequation:
+    # 2**(tree_height - 1) * cluster_size < num_labels <= 2**tree_height * cluster_size
+    height = int(np.ceil(np.log2(num_labels / cluster_size)))
+    logger.info(f"Labels will be grouped into {2**height} clusters")
+
+    output_dir.mkdir(parents=True, exist_ok=True)
+
+    # the normalized representations of the relationship between labels and texts
+    label_repr = normalize(sparse_y.T @ csc_matrix(sparse_x))
+
+    # clustering process
+    clusters = [np.arange(num_labels)]
+    for _ in range(height):
+        next_clusters = []
+        for cluster in clusters:
+            next_clusters.extend(_split_cluster(cluster, label_repr))
+        clusters = next_clusters
+        logger.info(f"Having grouped {len(clusters)} clusters")
+
+    np.save(cluster_path, np.asarray(clusters, dtype=object))
+    logger.info(f"Label clustering finished. Saving results to {repr(cluster_path)}")
+
+
+def _split_cluster(cluster: ndarray, label_repr: csr_matrix) -> tuple[ndarray, ndarray]:
+    """A variant of KMeans implemented in AttentionXML. The cluster is partitioned into two groups, each with
+    approximately equal size. Its main differences with the KMeans algorithm in scikit-learn are:
+    1. the distance metric is cosine similarity.
+    2. the end-of-loop criterion is the difference between the new and old average in-cluster distances to centroids.
+
+    Args:
+        cluster: a subset of labels
+        label_repr: the normalized representations of the relationship between labels and texts
+    """
+    tol = 1e-4
+
+    # the normalized label representations corresponding to the cluster
+    tgt_repr = label_repr[cluster]
+
+    # the number of labels in the cluster
+    n = len(cluster)
+
+    # Randomly choose two points as initial centroids and obtain their label representations
+    centroids = tgt_repr[np.random.choice(n, size=2, replace=False)].toarray()
+
+    # Initialize distances (cosine similarity)
+    # The cosine similarity always belongs to the interval [-1, 1]
+    old_dist = -2.0
+    new_dist = -1.0
+
+    # "c" denotes clusters
+    c0_idx = None
+    c1_idx = None
+
+    while new_dist - old_dist >= tol:
+        # each point's distances (cosine similarity) to the two centroids
+        # tgs_repr and centroids.T have been normalized
+        dist = tgt_repr @ centroids.T  # shape: (n, 2)
```

Comment:
can do 
```suggestion
        dist = label_repr[cluster] @ centroids.T  # shape: (n, 2)
```
here or we give label_repr[cluster] as input of this subroutine